src/fu-device-list.c

```diff
@@ -916,7 +916,7 @@ fu_device_list_wait_for_replug(FuDeviceList *self, GError **error)
 		g_main_context_iteration(NULL, FALSE);
 		devices_wfr_tmp = fu_device_list_get_wait_for_replug(self);
 		if (devices_wfr_tmp->len == 0)
-			break;
+			continue;
```

Comment:
this is not fine; i'll break lots of other plugins. The existing code will exit the loop when no devices are in the WAIT_FOR_REPLUG state, this change make the loop wait the entire remove_delay which for some other devices is a looong time. Can you revert this one and then attach a `-v -v` log showing the failure please. If there has to be a delay before accessing the device after replug for some reason this can be done in the plugin.

Reply:
Could you to how to fix the replug error? When re-enumerate our device, it usually fail~

Reply:
The error log:
09:04:02.282 FuDeviceList         waiting 5000ms for replug 
09:04:03.318 FuEngine             auto-removing backend device DISUM DFU Update V:0001 [b0a78eb71f4eeea7df8fb114522556ba8ce22074] 
09:04:03.318 FuDeviceList         waiting 5000ms for DISUM DFU Update V:0001 device removal 
09:04:03.318 FuEngine             resetting update motd timeout 
failed to get device before update reload: failed to wait for detach replug: device b0a78eb71f4eeea7df8fb114522556ba8ce22074 did not come back

It shows that though the device timer out length is 5000ms, the fwupd just waited 30ms.

Reply:
>  waiting 5000ms for DISUM DFU Update V:0001 device removal

So this means fwupd detected the hardware was "unplugged"

> failed to get device before update reload: failed to wait for detach replug

This is saying: 

 * detach put the device into bootloader mode
 * we waited for the device to re-enumerate
 * we installed the firmware
 * attach put the device into runtime mode
 * we waited for the device to re-enumerate
 * we read the new firmware version <- **failed before it got here**

the only thing that looks a bit odd is the handling of IS_BOOTLOADER in plugins/fpc/fu-fpc-device.c -- normally the bootloader flag is set or cleared in setup() by asking the hardware (like fu_fpc_device_setup_mode() seems to be doing) -- rather than setting it manually in attach() and detach(). Is that something we could do? I'm wondering if you're hitting the

	if (!fu_device_has_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
		g_debug("already in runtime mode, skipping");
		return TRUE;
	}

...and therefore not actually calling FPC_CMD_DFU_DETACH. If you could attach the full log with `-v -v` (warning -- it's a lot of lines!) then we can see what's happening. I think removing the manual add/remove of `FWUPD_DEVICE_FLAG_IS_BOOTLOADER` in fu_fpc_device_attach() and fu_fpc_device_detach() would make sense too.

Reply:
The command line is:
sudo ./fwupdtool --plugin-whitelist fpc --**verbose** install-blob disum.signed b0a78eb71f4eeea7df8fb114522556ba8ce22074

There is no already in runtime mode, skipping in the console output when error happens.

The log is :
Waiting…: 33%
Writing…: 33%
Writing…: 36%
Writing…: 40%
Writing…: 43%
Writing…: 48%
Writing…: 52%
Writing…: 55%
Writing…: 59%
Writing…: 63%
Writing…: 66%
Writing…: 70%
Writing…: 73%
Writing…: 77%
Writing…: 81%
Writing…: 85%
Writing…: 88%
Writing…: 93%
Waiting…: 93%
10:38:34.111 FuEngine             writing motd target /var/local/cache/fwupd/motd.d/85-fwupd
Waiting…: 96%
10:38:34.399 FuProgress           raw timing data was { 0.001, 3.930, 1.114 } -- steps were set as [ 1 95 5 ] but should have been [ 0 78 22 ] at ../plugins/fpc/fu-fpc-device.c:438
10:38:34.399 FuHistory            modifying device DISUM DFU Update V:0001 [b0a78eb71f4eeea7df8fb114522556ba8ce22074]
Restarting device…: 96%
10:38:34.400 FuEngine             install phase now attach
10:38:34.400 FuDeviceList         no replug or re-enumerate required
10:38:34.400 FuEngine             attach -> FuFpcDevice:
  DeviceId:             b0a78eb71f4eeea7df8fb114522556ba8ce22074
  Name:                 DISUM DFU Update V:0001
  Guid:                 bcb666d2-d13b-5295-b6ef-32a3f4ba8737 ← USB\VID_10A5&PID_9800
  Guid:                 7651ef86-f022-5508-8bcb-7cfa74cde98b ← USB\VID_10A5&PID_9800&REV_0000
  Serial:               6&3aa7951d&0&0001
  Summary:              FPC fingerprint sensor
  Plugin:               fpc
  Protocol:             com.fingerprints.dfupc
  Flags:                supported|registered|use-runtime-version|is-bootloader|updatable-hidden|signed-payload
  Problems:             update-in-process
  Vendor:               FPC
  VendorId:             USB:0x10A5
  Version:              27.26.23.18
  VersionFormat:        quad
  VersionRaw:           0x1b1a1712
  InstallDuration:      15
  Created:              2023-06-28
  UpdateState:          success
  Guid[quirks]:         dd4a4bc8-89bb-5453-9c8d-70ccd761b725 ← USB\VID_10A5
  Guid[quirks]:         c418fefd-0db6-5152-a4da-7a7527183cd1 ← USB\CLASS_FE
  Guid[quirks]:         ab182806-032b-5b7b-994f-b0713031d2b6 ← USB\CLASS_FE&SUBCLASS_01
  Guid[quirks]:         a55d6961-d2c3-5659-a26c-528521d001cd ← USB\CLASS_FE&SUBCLASS_01&PROT_02
  PhysicalId:           usb:01:00:01
  RemoveDelay:          5000
  AcquiesceDelay:       2500
  CustomFlags:          moh-device,rts,lenfy
  FirmwareSizeMin:      65536
  FirmwareSizeMax:      409600
  Order:                0
  PossiblePlugin:       fpc
  PossiblePlugin:       dfu
  ParentPhysicalIds:    usb:01:00,usb:01
  PrivateFlags:         moh-device,rts,lenfy
  Inhibit:              [update-in-process] An update is in progress
  Max block size:       0x1000
  LegacyDfu:            false
  MocDevice:            false
  RtsDevice:            true

10:38:34.400 FuPlugin             attach(fpc)
Waiting…: 96%
Waiting…: 97%
10:38:34.407 FuEngine             install phase now reload
10:38:34.407 FuDeviceList         waiting 5000ms for replug
10:38:35.443 FuEngine             auto-removing backend device DISUM DFU Update V:0001 [b0a78eb71f4eeea7df8fb114522556ba8ce22074]
10:38:35.443 FuDeviceList         waiting 5000ms for DISUM DFU Update V:0001 device removal
10:38:35.443 FuEngine             resetting update motd timeout
failed to get device before update reload: failed to wait for detach replug: device b0a78eb71f4eeea7df8fb114522556ba8ce22074 did not come back
10:38:45.992 FuPluginTest         destroy
10:38:45.993 FuProgress           raw timing data was { 0.000, 0.000, 0.004, 0.000 } -- steps were set as [ 1 1 3 95 ] but should have been [ 6 1 92 1 ] at ../libfwupdplugin/fu-context.c:810
10:38:45.996 FuProgress           raw timing data was { 0.001, 5.817, 0.007, 0.000 } -- steps were set as [ 2 94 2 2 ] but should have been [ 0 100 0 0 ] at ../plugins/fpc/fu-fpc-device.c:541
10:38:45.996 FuProgress           raw timing data was { 0.001, 3.930, 1.114 } -- steps were set as [ 1 95 5 ] but should have been [ 0 78 22 ] at ../plugins/fpc/fu-fpc-device.c:438
fpc@2004:~/code/fwupd/build/src$

Reply:
I have removed fu_device_remove_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)  in fu_fpc_device_attach.
And Ihave removed fu_device_add_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER) in fu_fpc_device_detach.
The  timeout problem still exists.

Reply:
The full log above is here:
[fulllog.txt](https://github.com/fwupd/fwupd/files/11933727/fulllog.txt)

Reply:
@jimzhang2 is that the full` --verbose --verbose` (you have to specify it twice) output?

Reply:
Also, if you use the daemon `sudo ./src/fwupd --verbose --verbose` does the device come back after 10 seconds? I'd be surprised if rebooting the device took that long. The extra verbose output would be super helpful. Also, is there any way I can test this on actual hardware locally? i.e. Could I get hardware with the affected problem so that we can add it to the test matrix?

Reply:
Got what you mean, I will add --verbose twice and add the log here

Reply:
> @jimzhang2 is that the full` --verbose --verbose` (you have to specify it twice) output?

No. That is just one --verbose in the log.

Reply:
> Also, if you use the daemon `sudo ./src/fwupd --verbose --verbose` does the device come back after 10 seconds? I'd be surprised if rebooting the device took that long. The extra verbose output would be super helpful. Also, is there any way I can test this on actual hardware locally? i.e. Could I get hardware with the affected problem so that we can add it to the test matrix?
Means you need the device to reproduce the error？

Reply:
> Also, if you use the daemon `sudo ./src/fwupd --verbose --verbose` does the device come back after 10 seconds? I'd be surprised if rebooting the device took that long. The extra verbose output would be super helpful. Also, is there any way I can test this on actual hardware locally? i.e. Could I get hardware with the affected problem so that we can add it to the test matrix?

Means you need the device to reproduce the error？

Reply:
command line :
sudo ./fwupdtool --plugin-whitelist fpc **--verbose --verbose** install-blob disum.signed b0a78eb71f4eeea7df8fb114522556ba8ce22074
error log :
[fulllog2.txt](https://github.com/fwupd/fwupd/files/11944853/fulllog2.txt)

Reply:
I Could not run ./src/fwupd --verbose --verbose before infinite testing loop.
Because the fwupd server will reject the repeat firmware download request.
So, the only thing I can do is  run 
sudo ./fwupdtool --plugin-whitelist fpc --verbose  --verbose install-blob disum_14.signed 0d5d05911800242bb1f35287012cdcbd9b381148
in the infinite  testing loop.
After the break->continue, the fwupd couldn't run normally in ubuntu22.04. It's weird.
But this change is OK in ubuntu20.04.

Reply:
Urgh, I get `Not Found` when downloading fulllog3.txt -- can you email me it please.

Reply:
We can't merge the `devices_wfr` bit.